src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentRepositoryBase.cs

```diff
@@ -1083,37 +1083,32 @@ protected virtual int GetReservedId(Guid uniqueId)
         protected void PersistRelations(TEntity entity)
         {
             // Get all references from our core built in DataEditors/Property Editors
-            // Along with seeing if deverlopers want to collect additional references from the DataValueReferenceFactories collection
-            var trackedRelations = new List<UmbracoEntityReference>();
-            trackedRelations.AddRange(_dataValueReferenceFactories.GetAllReferences(entity.Properties, PropertyEditors));
-
-            var relationTypeAliases = GetAutomaticRelationTypesAliases(entity.Properties, PropertyEditors).ToArray();
+            // Along with seeing if developers want to collect additional references from the DataValueReferenceFactories collection
+            var trackedRelations = _dataValueReferenceFactories.GetAllReferences(entity.Properties, PropertyEditors);
 
             // First delete all auto-relations for this entity
+            var relationTypeAliases = _dataValueReferenceFactories.GetAutomaticRelationTypesAliases(entity.Properties, PropertyEditors).ToArray();
             RelationRepository.DeleteByParent(entity.Id, relationTypeAliases);
 
             if (trackedRelations.Count == 0)
             {
                 return;
             }
 
-            trackedRelations = trackedRelations.Distinct().ToList();
-            var udiToGuids = trackedRelations.Select(x => x.Udi as GuidUdi)
-                .ToDictionary(x => (Udi)x!, x => x!.Guid);
+            var udiToGuids = trackedRelations.Select(x => x.Udi as GuidUdi).WhereNotNull().ToDictionary(x => (Udi)x, x => x.Guid);
 
             // lookup in the DB all INT ids for the GUIDs and chuck into a dictionary
-            var keyToIds = Database.Fetch<NodeIdKey>(Sql()
+            var keyToIds = Database.FetchByGroups<NodeIdKey, Guid>(udiToGuids.Values, Constants.Sql.MaxParameterCount, guids => Sql()
                 .Select<NodeDto>(x => x.NodeId, x => x.UniqueId)
                 .From<NodeDto>()
-                .WhereIn<NodeDto>(x => x.UniqueId, udiToGuids.Values))
+                .WhereIn<NodeDto>(x => x.UniqueId, guids))
                 .ToDictionary(x => x.UniqueId, x => x.NodeId);
 
-            var allRelationTypes = RelationTypeRepository.GetMany(Array.Empty<int>())?
-                .ToDictionary(x => x.Alias, x => x);
+            var allRelationTypes = RelationTypeRepository.GetMany(Array.Empty<int>()).ToDictionary(x => x.Alias, x => x);
 
             IEnumerable<ReadOnlyRelation> toSave = trackedRelations.Select(rel =>
                 {
-                    if (allRelationTypes is null || !allRelationTypes.TryGetValue(rel.RelationTypeAlias, out IRelationType? relationType))
+                    if (!allRelationTypes.TryGetValue(rel.RelationTypeAlias, out IRelationType? relationType))
```

Comment:
Should this also check whether the relation type returned by the reference is actually marked as an automatic relation using `relationTypeAliases.Contains(rel.RelationTypeAlias)`?

If this isn't the case, the relation that's added here won't get removed (since we're only removing the auto-relations for this entity at the top of the method)...

Reply:
PR https://github.com/umbraco/Umbraco-CMS/pull/15160 contains a fix that requires an automatic relation type, besides also being more graceful when it doesn't exist (by logging, instead of throwing an exception)...